Fix RevealsMap's effect not being removed for non-owners.

MustaphaTR opened the pull request:
Fixes #19147.

MustaphaTR committed:
Fix RevealsMap's effect not being removed for non-owners.

phrohdoh commented on OpenRA.Mods.Common/Traits/RevealsMap.cs:
```diff
@@ -44,7 +44,13 @@ protected void AddCellsToPlayerShroud(Actor self, Player p, PPos[] uv)
 		}
 
-		protected void RemoveCellsFromPlayerShroud(Actor self, Player p) { p.Shroud.RemoveSource(this); }
+		protected void RemoveCellsFromPlayerShroud(Actor self, Player p)
```
IMO methods should do exactly what they say on the tin and callers should only call methods with the data they want modified, so this'd only be invoked for the players that should actually have cells removed from their shroud.

MustaphaTR replied:
I thought about that too, but wasn't sure because check was in AddCellsToPlayerShroud for the adding part. I'll go change it for both.

phrohdoh commented on OpenRA.Mods.Common/Traits/RevealsMap.cs:
```diff
@@ -36,13 +36,7 @@ public RevealsMap(RevealsMapInfo info)
-
-			p.Shroud.AddSource(this, type, uv);
-		}
+		protected void AddCellsToPlayerShroud(Actor self, Player p, PPos[] uv) { p.Shroud.AddSource(this, type, uv); }
```
Oh, sorry I had not realized `AddCellsToPlayerShroud` existed and established the pattern of handling relationships internally. :(

Inq8 commented:
The first fix worked, the additional style/method change broke the trait completely.

MustaphaTR commented:
> the additional style/method change broke the trait completely.

Those returns should've been continue in the foreach loop. Tho before trying to change it again, i wanna be sure if this way is what we want or the previous want for the style. I'm also thinking that the check could be in its own function so it is not repeated 4 times.

pchote commented:
Why not
```csharp
foreach (var player in self.World.Players)
	if (Info.ValidRelationships.HasRelationship(self.Owner.RelationshipWith(player)))
		RemoveCellsFromPlayerShroud(self, player);
```
?

pchote commented:
I understand that we need to be looping over all the players in this trait and applying the `HasRelationship` check on both addition and removal. But there are several things I don't understand about the approach here...
* Why has the check been pulled out of `AddCellsToPlayerShroud` and duplicated across all the loops?
* Why don't we just add the check to `RemoveCellsFromPlayerShroud` instead?
* Isn't this also going to be a problem for `RevealsShroud` and `CreatesShroud`?

pchote commented:
Ping @MustaphaTR

MustaphaTR commented:
Rebased and updated, after looking at this and RevealsShroud again, i have realised that we don't need to do Relationship checks on RemoveCellsFromPlayerShroud. Any source that is added should be removed when actor dies/disposes or trait gets disabled and any that isn't added will be cought by the TryGetValue in Shroud.RemoveSource and it will return without doing anything. So this now only adds foreach loops for Removals under Disposed/Killed/TraitDisabled so it runs for every player, not just owner.

Mailaender reviewed (dismissed):
Works as promised.

RoosterDragon reviewed (dismissed):
Change looks good for solving the identified issue.

I have an adjacent thought which I haven't tested so not a blocker. What happens if the owner of the actor, or the relationships are changed? Would that cause any issues? Would the cells effectively need to be all revoked and then granted fresh in that scenario so things still match up?

MustaphaTR commented:
Relationships can't change, OpenRA doesn't support that, and for owner change, in my usecase the trait is on a dummy actor so it can't change owners but in general seems like a legitimate concern. AffectShroud has checks of AddedTo/RemovedFrom world, which i think handles owner change too, and it is another thing this trait doesn't check for, so more checks may be needed here. I should test.

MustaphaTR commented:
Updated and added INotifyOwnerChanged to reset the shroud when the owner changes. I decided this doesn't need Added/RemovedFromWorld checks, because it reveals the whole map so its position doesn't matter. As such i don't see any reason why this trait should not work when the actor is removed from the world.

RoosterDragon commented on OpenRA.Mods.Common/Traits/RevealsMap.cs:
```diff
@@ -51,14 +51,28 @@ protected PPos[] ProjectedCells(Actor self)
+				foreach (var player in self.World.Players)
+				{
+					RemoveCellsFromPlayerShroud(self, player);
+					AddCellsToPlayerShroud(self, player, ProjectedCells(self));
```
For performance, do `var cells = ProjectedCells(self);` outside the loop, do the same in `TraitEnabled`.

See a similar pattern in `AffectsShroud.UpdateShroudCells` for an example.

MustaphaTR replied:
Done.

Mailaender commented:
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/fea425e2993120d25ded7ac36444962e5af59a2f...1f36ea347883280c76517e2fe2f0266d0cd911ee)
